pennylane/transforms/batch_partial.py

```diff
@@ -0,0 +1,193 @@
+        an initial batch dimension.
+
+    **Example**
+
+    Consider the following circuit:
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=1)
+            return qml.expval(qml.PauliZ(wires=0) @ qml.PauliZ(wires=1))
+
+    The ``qml.batch_partial`` decorator allows us to create a partially evaluated
+    function that wraps the QNode. For example,
+
+    >>> y = np.array(0.2)
+    >>> batched_partial_circuit = qml.batch_partial(circuit, y=y)
+
+    The unevaluated arguments of the resulting function must now have a batch
+    dimension, and the output of the function also has a batch dimension:
+
+    >>> batch_size = 4
+    >>> x = np.linspace(0.1, 0.5, batch_size)
+    >>> batched_partial_circuit(x)
+    tensor([0.97517033, 0.95350781, 0.91491915, 0.86008934], requires_grad=True)
+
+    Jacobians can be computed for the arguments of the wrapper function, but
+    not for any partially evaluated arguments passed to ``qml.batch_partial``:
+
+    >>> qml.jacobian(batched_partial_circuit)(x)
+    array([[-0.0978434 ,  0.        ,  0.        ,  0.        ],
+           [ 0.        , -0.22661276,  0.        ,  0.        ],
+           [ 0.        ,  0.        , -0.35135943,  0.        ],
+           [ 0.        ,  0.        ,  0.        , -0.46986895]])
+
+    The same ``qml.batch_partial`` function can also be used to replace arguments
+    of a QNode with functions, and calling the wrapper would evaluate
+    those functions and pass the results into the QNode. For example,
+
+    >>> x = np.array(0.1)
+    >>> y_fn = lambda y0: y0 * 0.2 + 0.3
+    >>> batched_lambda_circuit = qml.batch_partial(circuit, x=x, y=y_fn)
+
+    The wrapped function ``batched_lambda_circuit`` also expects arguments to
+    have an initial batch dimension:
+
+    >>> batch_size = 4
+    >>> y0 = np.linspace(0.5, 2, batch_size)
+    >>> batched_lambda_circuit(y0)
+    tensor([0.91645953, 0.8731983 , 0.82121237, 0.76102116], requires_grad=True)
+
+    Jacobians can be computed in this scenario as well:
+
+    >>> qml.jacobian(batched_lambda_circuit)(y0)
+    array([[-0.07749457,  0.        ,  0.        ,  0.        ],
+           [ 0.        , -0.09540608,  0.        ,  0.        ],
+           [ 0.        ,  0.        , -0.11236432,  0.        ],
+           [ 0.        ,  0.        ,  0.        , -0.12819986]])
+    """
+    qnode = qml.batch_params(qnode, all_operations=all_operations)
+
+    preprocess = {} if preprocess is None else preprocess
+
+    # store whether this decorator is being used as a pure
+    # analog of functools.partial, or whether it is used to
+    # wrap a QNode in a more complex lambda statement
+    is_partial = preprocess == {}
+
+    # determine which arguments need to be stacked along the batch dimension
+    to_stack = []
```

Comment:
Why is this bookkeeping for future stacking done?

Reply:
It only makes sense to add a batch dimension to tensors. As you said before, the user can pass other types of objects (e.g. callables), and those don't need an extra batch dimension.